### ISSUE TITLE
$300 - Linux Wayland support 

### ISSUE THREAD

**Issue by bndlfm** [mediar-ai/screenpipe#63]
I encountered this error on NixOS with Nvidia proprietary drivers, I suspect it has to do with Wayland's security model around screen capture. My window manager is hyprland, kernel 6.10. I'm not sure it matters but I'm on a triple monitor setup.

`[2024-07-24T07:44:25Z INFO  screenpipe_server::video] Starting new video capture`
`[2024-07-24T07:44:25Z INFO  screenpipe_server::video] Started capture thread`
`thread 'tokio-runtime-worker' panicked at /home/runner/work/screen-pipe/screen-pipe/screenpipe-vision/src/utils.rs:130:42:
called 'Result::unwrap()' on an 'Err' value: Error("Screenshot failed or canceled")`
`note: run with 'RUST_BACKTRACE=1' environment variable to display a backtrace`

RUST_BACKTRACE=1:
`[2024-07-24T08:21:30Z INFO  screenpipe_server::video] Starting new video capture`
`[2024-07-24T08:21:30Z INFO  screenpipe_server::video] Started capture thread`
`*** bit out of range 0 - FD_SETSIZE on fd_set ***: terminated`

RUST_BACKTRACE=full
`[2024-07-24T08:39:53Z INFO  screenpipe_server::video] Started capture thread
thread 'tokio-runtime-worker' panicked at /home/runner/work/screen-pipe/screen-pipe/screenpipe-vision/src/utils.rs:130:42:
called 'Result::unwrap()' on an 'Err' value: Error("Screenshot failed or canceled")
stack backtrace:
   0:     0x562dcf09f855 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h1e1a1972118942ad
   1:     0x562dcf0ce13b - core::fmt::write::hc090a2ffd6b28c4a
   2:     0x562dcf09b24f - std::io::Write::write_fmt::h8898bac6ff039a23
   3:     0x562dcf09f62e - std::sys_common::backtrace::print::ha96650907276675e
   4:     0x562dcf0a0a99 - std::panicking::default_hook::{{closure}}::h215c2a0a8346e0e0
   5:     0x562dcf0a07dd - std::panicking::default_hook::h207342be97478370
   6:     0x562dcf0a0f33 - std::panicking::rust_panic_with_hook::hac8bdceee1e4fe2c
   7:     0x562dcf0a0e14 - std::panicking::begin_panic_handler::{{closure}}::h00d785e82757ce3c
   8:     0x562dcf09fd19 - std::sys_common::backtrace::__rust_end_short_backtrace::h1628d957bcd06996
   9:     0x562dcf0a0b47 - rust_begin_unwind
  10:     0x562dce2f1173 - core::panicking::panic_fmt::hdc63834ffaaefae5
  11:     0x562dce2f1626 - core::result::unwrap_failed::h82b551e0ff2b2176
  12:     0x562dce51d187 - screenpipe_server::video::VideoCapture::new::{{closure}}::h5000302bf0d7b53e
  13:     0x562dce536b1c - tokio::runtime::task::core::Core<T,S>::poll::h7296c4de051fa343
  14:     0x562dce352fa3 - tokio::runtime::task::harness::Harness<T,S>::poll::ha73eaf8f874d7aac
  15:     0x562dcefef64c - tokio::runtime::scheduler::multi_thread::worker::Context::run_task::h0a0e7670deb7415f
  16:     0x562dcefeeb0f - tokio::runtime::scheduler::multi_thread::worker::Context::run::h92da23e27729d718
  17:     0x562dceff9511 - tokio::runtime::context::runtime::enter_runtime::hca1473ffc8d65bdb
  18:     0x562dcefedd08 - tokio::runtime::scheduler::multi_thread::worker::run::h2eecb704a8589a50
  19:     0x562dceffb2c3 - <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll::hddd2d0453613d72b
  20:     0x562dcf00b558 - tokio::runtime::task::core::Core<T,S>::poll::hd3f247171015257b
  21:     0x562dcefe87a2 - tokio::runtime::task::harness::Harness<T,S>::poll::hb96f51c170b0f956
  22:     0x562dcf006d88 - tokio::runtime::blocking::pool::Inner::run::hff1cec3f73a737b7
  23:     0x562dcf00823c - std::sys_common::backtrace::__rust_begin_short_backtrace::hc65633ab330fbee4
  24:     0x562dcf00a932 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h09007c5a4f50fe68
  25:     0x562dcf0a656b - std::sys::pal::unix::thread::Thread::new::thread_start::h522bc89a54da820a
  26:     0x7f7982299a42 - start_thread
  27:     0x7f798231905c - __GI___clone3
  28:                0x0 - <unknown>`

EDIT: Just tested and it appears to work fine on x11, screen cap/recording definitely works.

**Comment by bndlfm** [comment]
I actually checked xcap but I wasn't sure since it implies partial support. I will try to build it with CUDA and report back.

As far as audio transcription without CUDA it works but interestingly it seems to repeat a lot.

![2024-07-24-144256_hyprshot](https://github.com/user-attachments/assets/13881bcd-3fad-47e9-bcd5-7579860dbf6a)

**Comment by louis030195** [comment]
@bndlfm thanks!

yes that's how whisper model act, we're about to push some heuristics to filter this that could easily fix this

if you query properly this noise shouldn't be too present atm also

eager to hear about cuda working

**Comment by louis030195** [comment]
@bndlfm also would love a contribution using https://github.com/helmerapp/scap instead of xcap which should have more compatibility (also using more modern Apple's API)

rough guess it's 10 lines of code to change

🙏

**Comment by bndlfm** [comment]
I'll have a look at it, but I must admit I'm not the best at coding. I was able to build screenpipe w/ cuda in a dev shell on NixOS with this flake (in case someone stumbles upon this issue who uses nix): 
``{
  inputs = {
    flake-utils.url = "github:numtide/flake-utils";
    naersk.url = "github:nix-community/naersk";
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
  };

  outputs = { self, flake-utils, naersk, nixpkgs }:
    flake-utils.lib.eachDefaultSystem (system:
      let
        pkgs = (import nixpkgs) {
          inherit system;
          config.allowUnfree = true;
          config.cudaSupport = true;
        };
        naersk' = pkgs.callPackage naersk {};

      in rec {
        # For `nix build` & `nix run`:
        defaultPackage = naersk'.buildPackage {
          src = ./.;
          nativeBuildInputs = with pkgs; [
            pkg-config
            cudaPackages.cudatoolkit
          ];
          buildInputs = with pkgs; [
            alsaLib
            llvmPackages.clang
            llvmPackages.libclang
            openssl
            xorg.libxcb
          ];
          propagatedBuildInputs = with pkgs; [
            ffmpeg
            tesseract
          ];
          LIBCLANG_PATH="${pkgs.llvmPackages.libclang.lib}/lib";
          CUDA_PATH="${pkgs.cudaPackages.cudatoolkit}";
          LD_LIBRARY_PATH="${pkgs.lib.makeLibraryPath [
            pkgs.cudaPackages.cudatoolkit
            pkgs.cudaPackages.cuda_cudart
          ]}";
        };

        # For `nix develop`:
        devShell = pkgs.mkShell.override { stdenv = if pkgs.config.cudaSupport then pkgs.cudaPackages.backendStdenv else pkgs.stdenv; } {
          packages = with pkgs; [ rustc cargo];
          nativeBuildInputs = with pkgs; [
            pkg-config
            cudaPackages.cudatoolkit
            cudaPackages.cuda_cudart
          ];
          buildInputs = with pkgs; [
            alsaLib
            llvmPackages.clang
            llvmPackages.libclang
            openssl
            xorg.libxcb
          ];
          propagatedBuildInputs = with pkgs; [
            ffmpeg
            tesseract
          ];
          shellHook = /*sh*/ ''
            export CUDA_PATH=${pkgs.cudaPackages.cudatoolkit}
            export LIBCLANG_PATH=${pkgs.llvmPackages.libclang.lib}/lib;
            export LD_LIBRARY_PATH=${pkgs.lib.makeLibraryPath [
              pkgs.cudaPackages.cudatoolkit
              pkgs.cudaPackages.cuda_cudart
            ]};
          '';
        };
      }
    );
}``

I haven't got ```nix build``` to work yet but ```nix develop && cargo build --release --features cuda``` works.

**Comment by wyyllou** [comment]
i got it to build with
cargo update
git add -f Cargo.lock
nix build

if you want to add the --features cuda you can probably add it to the `cargoOptions`

**Comment by bndlfm** [comment]
This builds perfectly but it fails to find [ "cuda" ] or [ "nvcuda"] according to ./screenpipe. These files don't seem to exist on NixOS, do they exist on normal linux as executables?

`{
  inputs = {
    flake-utils.url = "github:numtide/flake-utils";
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
  };

  outputs = { self, flake-utils, nixpkgs }:
    flake-utils.lib.eachDefaultSystem (system:
      let
        pkgs = (import nixpkgs) {
          inherit system;
          config.allowUnfree = true;
          config.cudaSupport = true;
        };
        cudaLibs = with pkgs.cudaPackages; [
          cudatoolkit
          cuda_cudart
          cuda_nvcc
          cuda_cccl
        ];
      in {
        defaultPackage = pkgs.rustPlatform.buildRustPackage.override {
          stdenv = if pkgs.config.cudaSupport
            then pkgs.cudaPackages.backendStdenv
            else pkgs.stdenv;
        } {
          pname = "screen-pipe";
          version = "0.1.0";
          src = ./.;
          buildFeatures = if pkgs.config.cudaSupport then [ "cuda" ] else [ ];
          cargoLock = {
            lockFile = ./Cargo.lock;
            outputHashes = {
              "cpal-0.15.3" = "sha256-eKn3tS5QuqbMTwnRAEybvbPZOiKiid7ghGztAmrs9fw=";
              "rusty-tesseract-1.1.10" = "sha256-XT74zGn+DetEBUujHm4Soe2iorQcIoUeZbscTv+64hw=";
            };
          };

          CUDA_COMPUTE_CAP = "89";
          CUDA_PATH = "${pkgs.cudaPackages.cudatoolkit}";
          LIBCLANG_PATH = "${pkgs.llvmPackages.libclang.lib}/lib";
          RUSTFLAGS = "-C link-arg=-lcuda -C link-arg=-lcudart";

          nativeBuildInputs = with pkgs; [
            pkg-config
            rustPlatform.bindgenHook
            autoPatchelfHook
            makeWrapper
          ] ++ cudaLibs;

          buildInputs = with pkgs; [
            alsaLib
            llvmPackages.clang
            llvmPackages.libclang
            openssl
            xorg.libxcb
          ] ++ cudaLibs;

          propagatedBuildInputs = with pkgs; [
            ffmpeg
            tesseract
          ] ++ cudaLibs;

          postPatch = let
            cudaLibPath = "${pkgs.cudaPackages.cudatoolkit}/lib:${pkgs.cudaPackages.cuda_cudart}/lib:${pkgs.linuxPackages.nvidia_x11}/lib";
          in ''
            substituteInPlace Cargo.toml \
              --replace 'cuda = []' 'cuda = ["${cudaLibPath}"]'
          '';

          postFixup = ''
            patchelf --add-rpath "${pkgs.lib.makeLibraryPath cudaLibs}" $out/bin/screenpipe
              wrapProgram $out/bin/screenpipe \
                --prefix LD_LIBRARY_PATH : "${pkgs.lib.makeLibraryPath cudaLibs}" \
                --prefix LD_PRELOAD : "${pkgs.lib.makeLibraryPath [ pkgs.cudaPackages.cuda_cudart ]}/libcudart.so"

            patchelf --add-rpath "${pkgs.lib.makeLibraryPath cudaLibs}" $out/bin/screenpipe-video
              wrapProgram $out/bin/screenpipe-video \
                --prefix LD_LIBRARY_PATH : "${pkgs.lib.makeLibraryPath cudaLibs}" \
                --prefix LD_PRELOAD : "${pkgs.lib.makeLibraryPath [ pkgs.cudaPackages.cuda_cudart ]}/libcudart.so"

            patchelf --add-rpath "${pkgs.lib.makeLibraryPath cudaLibs}" $out/bin/screenpipe-audio
              wrapProgram $out/bin/screenpipe-audio \
                --prefix LD_LIBRARY_PATH : "${pkgs.lib.makeLibraryPath cudaLibs}" \
                --prefix LD_PRELOAD : "${pkgs.lib.makeLibraryPath [ pkgs.cudaPackages.cuda_cudart ]}/libcudart.so"

            patchelf --add-rpath "${pkgs.lib.makeLibraryPath cudaLibs}" $out/bin/screenpipe-vision
              wrapProgram $out/bin/screenpipe-vision \
                --prefix LD_LIBRARY_PATH : "${pkgs.lib.makeLibraryPath cudaLibs}" \
                --prefix LD_PRELOAD : "${pkgs.lib.makeLibraryPath [ pkgs.cudaPackages.cuda_cudart ]}/libcudart.so"
          '';

          meta = {
            description = "screen-pipe";
            homepage = "https://github.com/louis030195/screen-pipe";
            license = pkgs.lib.licenses.mit;
            maintainers = with pkgs.lib.maintainers; [ ];
          };
        };



        devShell = pkgs.mkShell.override { stdenv = if pkgs.config.cudaSupport then pkgs.cudaPackages.backendStdenv else pkgs.stdenv; } {
          packages = with pkgs; [ rustc cargo];
          nativeBuildInputs = with pkgs; [
            pkg-config
            cudaPackages.cudatoolkit
            cudaPackages.cuda_cudart
          ];
          buildInputs = with pkgs; [
            alsaLib
            llvmPackages.clang
            llvmPackages.libclang
            openssl
            xorg.libxcb
          ];
          propagatedBuildInputs = with pkgs; [
            ffmpeg
            tesseract
          ];
          shellHook = /*sh*/ ''
            export CUDA_PATH=${pkgs.cudaPackages.cudatoolkit}
            export LD_LIBRARY_PATH=${pkgs.lib.makeLibraryPath [
              pkgs.cudaPackages.cudatoolkit
              pkgs.cudaPackages.cuda_cudart
            ]};
            export LIBCLANG_PATH=${pkgs.llvmPackages.libclang.lib}/lib;
          '';
        };
      }
    );
}
`

**Comment by louis030195** [comment]
/bounty 100

anyone wants to make screenpipe works on wayland? we don't have many wayland users but i'd be happy to support it!

definition of done
- macos & windows behaviour unchanged
- screen capture works on wayland 

suggestion to implement it:
- either make wayland work in https://github.com/nashaofu/xcap that we currently use
- or use https://github.com/CapSoftware/scap for linux only (and you likely have to implement app name and window name which are not implemented i think in scap)

**Comment by louis030195** [comment]
@fractalfir maybe interested? $300 bounty for this

**Comment by FractalFir** [comment]
Yeah, I could probably take a look. I may be a bit slower with response times.

**Comment by FractalFir** [comment]
I have built it on Fedora, running Gnome Wayland. It captures some screenshots just fine, other times, it freezes, and then segfaults. This is not something that happens on X11, so I will try to investigate that. 

```
you are using local processing. all your data stays on your computer.

2024-10-02T17:01:14.773225Z  INFO screenpipe_server::server: Server starting on 127.0.0.1:3030    
2024-10-02T17:01:18.915578Z  INFO screenpipe_server::video: Starting new video capture    
2024-10-02T17:01:18.915646Z  INFO screenpipe_server::video: Started capture thread    
[1]    47526 segmentation fault (core dumped)  cargo run --release --bin screenpipe

```

**Comment by FractalFir** [comment]
I can't reproduce the segfault in debug, I will try building release with debug symbols.

**Comment by FractalFir** [comment]
Well, I managed to get screenpipe to segfault under a debugger. Sadly, it is not of much help, since a lot of the variables seem to point to illegal locations.

```
[Detaching after vfork from child process 22395]

Thread 119 "tokio-runtime-w" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffbc600680 (LWP 22213)]
0x0000555558b2ece1 in socket_do_iteration (transport=<error reading variable: Cannot access memory at address 0x907fffffe8>, flags=<error reading variable: Cannot access memory at address 0x907fffffe4>, 
    timeout_milliseconds=<error reading variable: Cannot access memory at address 0x907fffffe0>) at ./vendor/dbus/dbus/dbus-transport-socket.c:1183
1183	      poll_res = _dbus_poll (&poll_fd, 1, poll_timeout);
(gdb) bt
#0  0x0000555558b2ece1 in socket_do_iteration (transport=<error reading variable: Cannot access memory at address 0x907fffffe8>, flags=<error reading variable: Cannot access memory at address 0x907fffffe4>, 
    timeout_milliseconds=<error reading variable: Cannot access memory at address 0x907fffffe0>) at ./vendor/dbus/dbus/dbus-transport-socket.c:1183
Backtrace stopped: Cannot access memory at address 0x9080000028
```

**Comment by FractalFir** [comment]
I also have discovered a **very severe** usability issue on Wayland Gnome. 

Gnome will flash the whole screen when a screenshot is taken. There is no way to disable this, besides compiling gnome from source. Because of this, screenpipe will flash the whole screen once a second, making it a potential source of seizures. It is also unusable in that state. 

I have included a video recording of this happening. 


**WARNING: THIS VIDEO CONTAINS FLASHING LIGHTS AND MIGHT CAUSE SEIZURES**

https://github.com/user-attachments/assets/fc7631aa-decd-470c-80c8-7692b1671dd8


I think screenpipe should warn people about this, and not launch on Wayland Gnome. 

In the future, some sort of workaround will be needed.

**Comment by tribhuwan-kumar** [comment]
@FractalFir  are you working on this ? or should I give it a try ??

**Comment by FractalFir** [comment]
> @FractalFir  are you working on this ? or should I give it a try ??

Well, I tried to solve it, and still plan to work on it, but I have to go to sleep now, and them have to do some other stuff tomorrow.


Looking at this, the issue seems to be a bit more complex than anticipated. 

There is the segfault issue(which is almost impossible to debug) the previous problems, and the issues with NixOS specifically.

**Comment by tribhuwan-kumar** [comment]
> > @FractalFir  are you working on this ? or should I give it a try ??
> 
> Well, I tried to solve it, and still plan to work on it, but I have to go to sleep now, and them have to do some other stuff tomorrow.
> 
> Looking at this, the issue seems to be a bit more complex than anticipated.
> 
> There is the segfault issue(which is almost impossible to debug) the previous problems, and the issues with NixOS specifically.

sure, take a test. we'll discuss about it later!!

**Comment by FractalFir** [comment]
I am now 99% sure the segfault I found is caused by `xcap`. `xcap` also appears to not be thread safe, and will sometimes hang when more than one thread is taking screenshots!

I would consider checking if this issue also happens on other platforms, since this could cause worker threads to stall one by one. This may not be something immediately obvious, since threads only sometimes hang. With many worker threads, this issue might only be noticeable after some time. 

```rust
use std::time::Instant;
use xcap::Monitor;

fn normalized(filename: &str) -> String {
    filename
        .replace("|", "")
        .replace("\\", "")
        .replace(":", "")
        .replace("/", "")
}
fn make_screenshots() {
    println!("Preparing to take a screenshot!");
    let monitors = Monitor::all().unwrap();
    for monitor in monitors {
        monitor.capture_image().unwrap();
    }
    println!("Taken a screenshot!");
}
fn main() {
    let start = Instant::now();
        // This thread will sometimes hang on Wayland.
    let handle = std::thread::spawn(|| {
        make_screenshots();
    });
    make_screenshots();

    handle.join();
}
```


Knowing all of that, I will:
1. Report the issue to the `xcap` maintainers.
2. Port `screenpipe` to use `scap` on Linux. 

In the future, once the `xcap` maintainers fix this bug, you could move back to using only `xcap`. So, I will try to keep the changes minimal.

**Comment by FractalFir** [comment]
@louis030195 I think I found the exact cause of this issue.

I have pinpointed the issue to a misuse of a single dbus function in xcap. 

`xcap` uses Connection::process without checking if another thread is already doing so,  which is not allowed.

https://docs.rs/dbus/latest/dbus/blocking/struct.Connection.html#method.process

This caused the message containing the screenshot to be lost when multiple threads are taking a screenshot at the same time.

I can provide a quick, albeit a bit inefficient solution to this problem, which will not require moving to `scap`.

We can simply use a Mutex to guarantee no 2 threads are taking a screenshot at the same time.

I think this is a decent solution - since it does not require adding another dependency, only required for supporting Wayland. What do you think?

Also, I think I see quite a few inefficiencies in `xcap` - it is hard-coded to wait at least a second before checking if a screenshot succeed. So, I might be able to make screenshots on Wayland in `xcap` quite a bit faster, without too much effort.

**Comment by FractalFir** [comment]
Ok, it seems there are even more issue with `xcap`. At this point, just switching to `scap` for Linux seems like the only option.

```
ead 101 "tokio-runtime-w" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffcd200680 (LWP 27902)]
0x0000555558b2691f in _dbus_transport_ref (transport=0x3b1b60003a6cc000) at ./vendor/dbus/dbus/dbus-transport.c:467
467	  transport->refcount += 1;
(gdb) bt
#0  0x0000555558b2691f in _dbus_transport_ref (transport=0x3b1b60003a6cc000) at ./vendor/dbus/dbus/dbus-transport.c:467
#1  0x0000555558b2dbc4 in do_io_error (transport=0x3b1b60003a6cc000) at ./vendor/dbus/dbus/dbus-transport-socket.c:238
#2  0x0000555558b2ed3a in socket_do_iteration (transport=0x3b1b60003a6cc000, flags=994467840, timeout_milliseconds=1002496000) at ./vendor/dbus/dbus/dbus-transport-socket.c:1204
Backtrace stopped: previous frame inner to this frame (corrupt stack?)

```

**Comment by FractalFir** [comment]
I think I got `xcap` working as expected. Screenshots are correctly made, but `screenpipe` now crashes in `rusty_tesseract`. I am not sure if it is `xcap` related, so I will continue to investigate.

**Comment by FractalFir** [comment]
Well, the issue turned out to be with `xcap` again. So, I will be sticking to `scap`.

I am partially done with moving linux builds to `scap`. I think I found a way to change the underlying library without touching too much other code.

**Comment by louis030195** [comment]
> Well, the issue turned out to be with `xcap` again. So, I will be sticking to `scap`.
> 
> I am partially done with moving linux builds to `scap`. I think I found a way to change the underlying library without touching too much other code.
> 
> However, I have to finish work for today, and will(hopefully) continue and fix this issue by tomorrow.


Great progress 🙏
Just make sure to have app_name and window_name available because we're highly relying on it (capturing each window and each app ...)

**Comment by FractalFir** [comment]
So, from taking a glance at the repo, it seems like scap does not support getting the names of windows. 
Because of that, I will try to get xcap working once again.

I think I reduced the severity of the issue, but I am a bit doubtful about the quality of `xcap` code. Debugging it has proven to be quite difficult.

**Comment by FractalFir** [comment]
I have a progress update: I managed to fix the `Error("Screenshot failed or canceled")` multi-threading bug in `xcap`!

https://github.com/nashaofu/xcap/commit/99af18b6ed6e449df0b28777717fcdac31307170

I will now open a pull request and try to get the changes upstream. 

There still are some other issues that I am trying to fix, but they may be only present on Gnome. I will need to do a bit more testing.

**Comment by FractalFir** [comment]
Well, from some pretty basic testing it *looks* like the `Error("Screenshot failed or canceled")` issue is no solved now.

However, the crashes are indeed caused by `xcap`, or, more specifically, by the `dbus` crate. 

It *looks* like this is a bug in `dbus` - `xcap` is using only safe APIs of `dbus`, so it should not be able to cause any crashes. 

I am investigating this issue, and I have reported it in the `dbus` repo.

**Comment by FractalFir** [comment]
Progress update: I think I got `xcap`(and thus `screenpipe`) to fully work on Wayland. 

I had run screenpipe for 30 minutes, and it did not crash, or encounter any errors. It made screenshots every second, for the whole 30 minutes. So, it seems to be fully working.

I have made  3 changes to `xcap`:
1. Updated dbus and changed it from vendored to system. This fixed the segault issue, but it sadly means that building screenpipe on Linux now requires libdbus to be installed. 
2. Added locks to ensure no 2 threads are taking screnshots at the same time.
3. Disabled  Gnome-specific screnshot code, which does not seem to be working ATM. This code seems to be a workaround for the flashing issue I encountered, but it does not work. 


Tomorrow, I will prepare my patches to `xcap` for a pull request. You will also be able to fork my repo with patches then, if you don't want to wait for the changes to be merged upstream.

**Comment by FractalFir** [comment]
@louis030195  I believe both the screenshot error  and segfaults should be fixed now, in [my PR](https://github.com/nashaofu/xcap/pull/156) to the `xcap` repo.  In the end, I only needed to change 2 files in `xcap`(Cargo.toml + wayland_capture.rs). The PR contains more detailed explanations of all the changes. If you have any questions about the PR, I'd be happy to answer.  


This seems like a full fix. I have been running screenpipe on Wayland for 30 + minutes, and there were no issues. I could leave it running for a few hours, but I don't think that I would encounter any more problems.

@bndlfm 

Could you check if the changes work for you too?

You can replace the `xcap` dependency in screenpipe-vision/Cargo.toml with

```
xcap = {git = "https://github.com/FractalFir/xcap.git", rev = "05186b191d800cd5f8d31b4723fca9b263eb5695"}
```
to try my PR.

**Comment by FractalFir** [comment]
@louis030195  the PR has now been merged into `xcap`. I don't know when they plan to  release a new version on `crates.io`, but you should be able to get a working version of `xcap` from their repo now.

**Comment by louis030195** [comment]
@FractalFir thanks! maybe send a PR to use their latest commit in screenpipe as dependency?

if possible I'd like to have someone else test out on wayland and confirm it works good then approve bounty and close issue

**Comment by FractalFir** [comment]
> @FractalFir thanks! maybe send a PR to use their latest commit in screenpipe as dependency?

I will open a PR, although AFAIK cargo does not support different crate sources for different targets. 

So, you'd have to update screen-pipe to the newest xcap commit on all targets, and I am not sure if you want to do that. So, you might need to wait for a new `xcap` release, if you want to use a different version of it on some OSs. 

> if possible I'd like to have someone else test out on wayland and confirm it works good

Of course, that is also what I wanted, and why I pinged the person who originally opened the issue.  Maybe somebody else who encountered problems on Wayland could test it out?  

Ideally, the patch should be tested by somebody who uses screenpipe a lot, since they would catch all potential issues.

I don't mind waiting a few days to get this tested - I would fell terrible if I got the bounty, only to learn that my patch is defective.
